### PR TITLE
Upgrading Elasticsearch to OTel version 1.0.2

### PR DESCRIPTION
**Pull request by ejsmith**
This PR upgrades the Elasticsearch instrumentation to OTel version 1.0.2 and serves as an example of how to get rid of `ActivitySourceAdapter` as well as use the new `AddLegacyActivity` method.

It also adds a new `OpenTelemetry.Contrib.Shared` project that contains all of the source shared from the main OTel .NET library that is necessary to create new instrumentations. The library is not meant to be directly referenced. Instead an instrumentation library should use the `IncludeSharedInstrumentationSource` MSBuild property in order to have the shared source files included in their project.

### CODE REVIEW

File: src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/OpenTelemetry.Contrib.Instrumentation.ElasticsearchClient.csproj
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/OpenTelemetry.Contrib.Instrumentation.ElasticsearchClient.csproj:9]: need to goto a version which has https://github.com/open-telemetry/opentelemetry-dotnet/pull/1836 changes , as thats the only which eliminates ActivitySourceAdapter.

Comment (cijothomas) [reply src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/OpenTelemetry.Contrib.Instrumentation.ElasticsearchClient.csproj:9]: we'll have it on myget soon, will validate and release to nuget after enough validations.

Comment (ejsmith) [reply src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/OpenTelemetry.Contrib.Instrumentation.ElasticsearchClient.csproj:9]: @cijothomas I'm not sure what that means. I see that you guys are working on removing `ActivitySourceAdapter`, but I'm not using it and I'm using a similar method as what is being done in that PR. I did notice the `AddLegacyOperationName` ability, but I don't see how you go about enriching the activities like I'm doing in this instrumentation when using that method.

Comment (ejsmith) [reply src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/OpenTelemetry.Contrib.Instrumentation.ElasticsearchClient.csproj:9]: Not to mention, that I see that you are STILL using a BUNCH of internal methods and utilities in your own instrumentations even after this https://github.com/open-telemetry/opentelemetry-dotnet/pull/1836 is merged. So I really don't see how that PR is going to help me much. If I just comment out the `InternalsVisibleTo` for your `GrpcNetClient` in this latest code there are still a bunch of things I can't do:

image

I really don't understand why you wouldn't get rid of those `InternalsVisibleTo` and stop cheating so you can see what it's like for contributors to try and create their own instrumentations.

Comment (cijothomas) [reply src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/OpenTelemetry.Contrib.Instrumentation.ElasticsearchClient.csproj:9]: > @cijothomas I'm not sure what that means. I see that you guys are working on removing `ActivitySourceAdapter`, but I'm not using it and I'm using a similar method as what is being done in that PR. I did notice the `AddLegacyOperationName` ability, but I don't see how you go about enriching the activities like I'm doing in this instrumentation when using that method.

With using `AddLegacyOperationName`, the activity created in ES library is not going through the samplers registered in the TracerProvider.

Comment (cijothomas) [reply src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/OpenTelemetry.Contrib.Instrumentation.ElasticsearchClient.csproj:9]: The internalsvisible is now only used for helpers. It can be copied by any project. (even in the main repo, they can be copied to individual instrumentations and remove internalvisible attribute. Feel free to submit a PR to restructure it. if you think its important.)

`ActivitySourceAdapter` is/was the only one which was specific to the OTel SDK, and folks outside that repo could not have just copied it over - it was done to unblock progress at the time, and now its being removed.

Comment (ejsmith) [reply src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/OpenTelemetry.Contrib.Instrumentation.ElasticsearchClient.csproj:9]: @cijothomas done https://github.com/open-telemetry/opentelemetry-dotnet/pull/1856

File: src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/TracerProviderBuilderExtensions.cs
Comment (ejsmith) [inline src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/TracerProviderBuilderExtensions.cs:48]: @cijothomas updated to use `AddLegacyActivity`. Am I doing it correctly?

Comment (cijothomas) [reply src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/TracerProviderBuilderExtensions.cs:48]: Looks correct. @utpilla and myself was planning to modify one of the instrumentations in -contrib to leverage the new ActivitySourceAdapter-free verion. Glad to see you are ahead of us!

@utpilla Please take a look. I'll do a detailed review Monday.

File: src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/Implementation/ElasticsearchRequestPipelineDiagnosticListener.cs
Comment (utpilla) [inline src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/Implementation/ElasticsearchRequestPipelineDiagnosticListener.cs:71]: If the intention is to change the `ActivitySource` of this activity to `ElasticsearchActivitySourceHelper.ActivitySourceName` and update the `ActivityKind` to `Client` then you should add these two lines here:
```
ActivityInstrumentationHelper.SetActivitySourceProperty(activity, ElasticsearchActivitySourceHelper.ActivitySourceName);
ActivityInstrumentationHelper.SetKindProperty(activity, ActivityKind.Client);
```

Comment (cijothomas) [reply src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/Implementation/ElasticsearchRequestPipelineDiagnosticListener.cs:71]: its there. lines 61,62 :)

Comment (utpilla) [reply src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/Implementation/ElasticsearchRequestPipelineDiagnosticListener.cs:71]: Oops! Never mind!

File: src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/TracerProviderBuilderExtensions.cs
Comment (utpilla) [reply src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/TracerProviderBuilderExtensions.cs:48]: This looks good!

Comment (cijothomas) [reply src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/TracerProviderBuilderExtensions.cs:48]: ok. We can modify the instrumentation unit test to validate that the activity has IsAllDataRequested/Sampled populated correctly based on the sampler. This will validate that those activities are fed through Samplers.

Comment (ejsmith) [reply src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/TracerProviderBuilderExtensions.cs:48]: Ok, I will give that a try.

Comment (ejsmith) [reply src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/TracerProviderBuilderExtensions.cs:48]: @cijothomas I added a sampler test and it does not appear to be working correctly. I think I may need `SuppressInstrumentationScope.IncrementIfTriggered` and `SuppressInstrumentationScope.DecrementIfTriggered` to be public so that their calls can be added back into `DiagnosticSourceListener`. I had to remove it from [here ](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/DiagnosticSourceInstrumentation/DiagnosticSourceListener.cs#L56) since it's currently internal.

File: test/OpenTelemetry.Contrib.Instrumentation.ElasticsearchClient.Tests/ElasticsearchClientTests.cs
Comment (cijothomas) [inline test/OpenTelemetry.Contrib.Instrumentation.ElasticsearchClient.Tests/ElasticsearchClientTests.cs:244]:
```suggestion
                .SetSampler(sampler)
```

File: test/OpenTelemetry.Contrib.Instrumentation.ElasticsearchClient.Tests/ElasticsearchClientTests.cs
Comment (cijothomas) [inline test/OpenTelemetry.Contrib.Instrumentation.ElasticsearchClient.Tests/ElasticsearchClientTests.cs:245]: until we get the newer build from main repo (the one with IncrementIfTriggered), can you try 
`.AddElasticsearchClientInstrumentation((opt) => opt.SuppressDownstreamInstrumentation = false)`

Comment (cijothomas) [reply test/OpenTelemetry.Contrib.Instrumentation.ElasticsearchClient.Tests/ElasticsearchClientTests.cs:245]: Once you do this (and fix the sampler -see above comment-), this test should pass...

File: src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/TracerProviderBuilderExtensions.cs
Comment (cijothomas) [reply src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/TracerProviderBuilderExtensions.cs:48]: I left couple of comments in the test code. If it still fails, can you verify if the activity.OperationName for the activity created by ES client is "Elasticsearch.Net.RequestPipeline.CallElasticsearch" itself? Not sure if its case sensitive.

File: test/OpenTelemetry.Contrib.Instrumentation.ElasticsearchClient.Tests/ElasticsearchClientTests.cs
Comment (ejsmith) [reply test/OpenTelemetry.Contrib.Instrumentation.ElasticsearchClient.Tests/ElasticsearchClientTests.cs:245]: I made both of those changes and it still fails. I had it using the `sampler` instance and then tried the `AlwaysOffSampler` and forgot to put it back. For some reason, it's not calling the `testActivityProcessor.StartAction` callback.

File: src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/TracerProviderBuilderExtensions.cs
Comment (CodeBlanch) [reply src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/TracerProviderBuilderExtensions.cs:48]: Are we still thinking about renaming to `AddLegacySource`?

Comment (utpilla) [reply src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/TracerProviderBuilderExtensions.cs:48]: @CodeBlanch Yeah we had decided to update the name of the extension method to `AddLegacySource` in a different PR after this [one](https://github.com/open-telemetry/opentelemetry-dotnet/pull/1836).

@ejsmith You have to pass `CallElasticsearch` to `AddLegacyActivity` as that's the OperationName of the activity created by the DiagnosticListener. You'll also have to update the method name to `AddLegacySource` when it gets updated in the new package.

File: test/OpenTelemetry.Contrib.Instrumentation.ElasticsearchClient.Tests/ElasticsearchClientTests.cs
Comment (utpilla) [reply test/OpenTelemetry.Contrib.Instrumentation.ElasticsearchClient.Tests/ElasticsearchClientTests.cs:245]: Please update the `AddElasticsearchClientInstrumentation` method in `TracerProviderBuilderExtensions.cs` to use `builder.AddLegacyActivity("CallElasticsearch");`.You'll have to update the other unit tests to expect the processor invocations to  equal 5 instead of 4 as the Processor.OnStart should also be called.

File: src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/TracerProviderBuilderExtensions.cs
Comment (ejsmith) [reply src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/TracerProviderBuilderExtensions.cs]: Thanks @utpilla that worked!

File: src/OpenTelemetry.Contrib.Shared/OpenTelemetry.Contrib.Shared.csproj
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Shared/OpenTelemetry.Contrib.Shared.csproj:7]: we probably don't need this or it should come from common props file.

File: src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/Implementation/ElasticsearchActivitySourceHelper.cs
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/Implementation/ElasticsearchActivitySourceHelper.cs:25]: nit: this class doesn't contain any helper method, so the old name "constants" was still good.

**Review by cijothomas — APPROVED** [review]
changes to move away from ActivitySourceAdapater is good and the tests confirm its working.

File: test/OpenTelemetry.Contrib.Instrumentation.ElasticsearchClient.Tests/ElasticsearchClientTests.cs
Comment (ejsmith) [reply test/OpenTelemetry.Contrib.Instrumentation.ElasticsearchClient.Tests/ElasticsearchClientTests.cs]: When is the method name change supposed to be merged into the main repo?

**Comment by cijothomas** [comment]
@ejsmith could you update the PR title and description to match the current state.
Changes look good to me, and can be merged once the CI is green.

**Comment by ejsmith** [comment]
@cijothomas I'm still waiting for resolution on https://github.com/open-telemetry/opentelemetry-dotnet/pull/1856 because I need access to `SuppressInstrumentationScope.IncrementIfTriggered` and `SuppressInstrumentationScope.DecrementIfTriggered` inside of the `DiagnosticSourceListener` class in order to make nested activity suppression work correctly.

File: src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/Implementation/ElasticsearchActivitySourceHelper.cs
Comment (ejsmith) [reply src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/Implementation/ElasticsearchActivitySourceHelper.cs:25]: I'm going to remove this class and add these constants to the `ElasticsearchRequestPipelineDiagnosticListener` class.

File: src/OpenTelemetry.Contrib.Shared/OpenTelemetry.Contrib.Shared.csproj
Comment (ejsmith) [reply src/OpenTelemetry.Contrib.Shared/OpenTelemetry.Contrib.Shared.csproj:7]: `Common.props` does not contain a `TargetFrameworks` element so this is required. Each instrumentation lib should probably choose it's own targets based on the library it's instrumenting.

**Comment by ejsmith** [comment]
@cijothomas is the title and description better now? All the checks should be passing now as well.

**Comment by alexvaluyskiy** [comment]
I’ve updated the instrumentation for MassTransit and EntityFrameworkCore. But I have to add the same shared files, like in this PR.
@ejsmith @cijothomas Is there any chance this PR will be merged soon? Or I should not wait and create my own PR?

**Comment by ejsmith** [comment]
Still waiting for resolution on https://github.com/open-telemetry/opentelemetry-dotnet/pull/1856 before this can be merged. I'm attempting to wait patiently. 😆

**Comment by cijothomas** [comment]
Thanks for waiting! 
We'll close on the dependent PRs today, and unblock all the PRs in this repo :)

File: src/OpenTelemetry.Contrib.Shared/DiagnosticSourceInstrumentation/DiagnosticSourceListener.cs
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Shared/DiagnosticSourceInstrumentation/DiagnosticSourceListener.cs:26]: remove these checks. (adapt https://github.com/open-telemetry/opentelemetry-dotnet/pull/1893)

**Comment by cijothomas** [comment]
Adapt changes like this https://github.com/open-telemetry/opentelemetry-dotnet/pull/1893 and this should unblock this PR including the supressdownstream functionality.

**Comment by ejsmith** [comment]
@cijothomas done

**Review by cijothomas — APPROVED** [review]

**Pull request merged**
